Bug:  `*` is treated as non-universal pattern

### Environment

Node version: v20.12.0
npm version: 10.5.0
Local ESLint version: v9.4.0
Global ESLint version: no
Operating System: windows


### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
export default [
  {
    // configuration for files in the root, typically config files that run in Node.js
    files: ['*'],
    languageOptions: {
      globals: {
        process: 'readonly',
      },
    },
    rules: {
      'no-undef': 2,
    },
  },
];

```
</details>

```js
process.exit(1);
```


### What did you expect to happen?

No errors when running `eslint .`

### What actually happened?

```
C:\projects\tmp\tmp\.gitignore
  1:1  error  'node_modules' is not defined  no-undef

C:\projects\tmp\tmp\package-lock.json
  2:9  error  Parsing error: Unexpected token :

C:\projects\tmp\tmp\package.json
  2:9  error  Parsing error: Unexpected token :
```

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/stackblitz-starters-5p7i7w?file=eslint.config.js

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

Patterns that end with `/*` are considered non-universal - they don't cause any matching files to be linted. Shouldn't the same apply to lone `*`?

The same could be argued for `**`, though maybe it makes sense to leave it as-is, as a special feature of `**` (otherwise, it would be the same as `**/*`).

--- a/package.json
+++ b/package.json
@@ -68,7 +68,7 @@
   "dependencies": {
     "@eslint-community/eslint-utils": "^4.2.0",
     "@eslint-community/regexpp": "^4.6.1",
-    "@eslint/config-array": "^0.15.1",
+    "@eslint/config-array": "^0.16.0",
     "@eslint/eslintrc": "^3.1.0",
     "@eslint/js": "9.4.0",
     "@humanwhocodes/module-importer": "^1.0.1",

--- a/tests/lib/eslint/eslint.js
+++ b/tests/lib/eslint/eslint.js
@@ -1305,6 +1305,22 @@ describe("ESLint", () => {
             assert.strictEqual(results[0].suppressedMessages.length, 0);
         });
 
+        // https://github.com/eslint/eslint/issues/18550
+        it("should skip files with non-standard extensions when they're matched only by a '*' files pattern", async () => {
+            eslint = new ESLint({
+                cwd: getFixturePath("files"),
+                overrideConfig: { files: ["*"] },
+                overrideConfigFile: true
+            });
+            const results = await eslint.lintFiles(["."]);
+
+            assert.strictEqual(results.length, 2);
+            assert(
+                results.every(result => /^\.[cm]?js$/u.test(path.extname(result.filePath))),
+                "File with a non-standard extension was linted"
+            );
+        });
+
         // https://github.com/eslint/eslint/issues/16413
         it("should find files and report zero messages when given a parent directory with a .js", async () => {
             eslint = new ESLint({

